File: Python/programmers/etc/level2/미로_탈출.py
from collections import deque

dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]

def solution(maps):
    row, col = len(maps), len(maps[0])
    visited = [[[-1]*2 for _ in range(col)] for _ in range(row)]
    for i in range(row):
        for j in range(col):
            if maps[i][j] == "S":
                sx, sy = i, j
            elif maps[i][j] == "L":
                lx, ly = i, j
            elif maps[i][j] == "E":
                ex, ey = i, j
    q = deque()
    q.append([sx, sy, 0])
    visited[sx][sy][0] = 0
    while q:
        x, y, state = q.popleft()
        for d in range(4):
            nx, ny = x+dx[d], y+dy[d]
            if not (0 <= nx < row and 0 <= ny < col and maps[nx][ny] != "X"):
                continue
            if (nx, ny) == (lx, ly) and state == 0 and visited[nx][ny][1] == -1:
                visited[nx][ny][1] = visited[x][y][0]+1
                q.append([nx, ny, 1])
            if visited[nx][ny][state] == -1:
                visited[nx][ny][state] = visited[x][y][state]+1
                q.append([nx, ny, state])
                
    return visited[ex][ey][1]


maps = ["SOOOL","XXXXO","OOOOO","OXXXX","OOOOE"]
print(solution(maps)) # 16

maps = ["LOOXS","OOOOX","OOOOO","OOOOO","EOOOO"]
print(solution(maps)) # -1

File: Java/programmers/etc/level2/미로탈출.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 미로탈출 {
    int[] dx = new int[]{-1, 1, 0, 0};
    int[] dy = new int[]{0, 0, -1, 1};
    
    public int solution(String[] tmp) {
        int row = tmp.length, col = tmp[0].length();
        String[][] maps = new String[row][col];
        for (int i=0; i<row; i++) {
            String[] str = tmp[i].split("");
            for (int j=0; j<col; j++) {
                maps[i][j] = str[j];
            }
        }
        
        int[][][] visited = new int[row][col][2];
        int sx = 0, sy = 0, lx = 0, ly = 0, ex = 0, ey = 0;
        for (int i=0; i<row; i++) {
            for (int j=0; j<col; j++) {
                if (maps[i][j].equals("S")) {
                    sx = i; sy = j;
                } else if (maps[i][j].equals("L")) {
                    lx = i; ly = j;
                } else if (maps[i][j].equals("E")) {
                    ex = i; ey = j;
                }
            }
        }
        
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{sx, sy, 0});
        visited[sx][sy][0] = 1;
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0], y = cur[1], state = cur[2];
            for (int d=0; d<4; d++) {
                int nx = x+dx[d], ny = y+dy[d];
                if (nx < 0 || nx >= row || ny < 0 || ny >= col || maps[nx][ny].equals("X")) continue;
                if ((nx == lx && ny == ly) && state == 0 && visited[nx][ny][1] == 0) {
                    visited[nx][ny][1] = visited[x][y][0]+1;
                    q.add(new int[]{nx, ny, 1});
                }
                if (visited[nx][ny][state] == 0) {
                    visited[nx][ny][state] = visited[x][y][state]+1;
                    q.add(new int[]{nx, ny, state});
                }
            }
        }
        return visited[ex][ey][1]-1;
    }
}
